solution/0300-0399/0323.Number of Connected Components in an Undirected Graph/README_EN.md

```diff
@@ -487,4 +487,47 @@ function countComponents(n: number, edges: number[][]): number {
 
 <!-- solution:end -->
 
+<!-- solution:start -->
+
+### Solution 3: BFS
+
+<!-- tabs:start -->
+
+#### TypeScript
+
+```ts
+function countComponents(n: number, edges: number[][]): number {
+    const g: number[][] = Array.from({ length: n }, () => []);
+    const vis = new Set<number>();
+    let ans = 0;
+
+    for (const [i, j] of edges) {
+        g[i].push(j);
+        g[j].push(i);
+    }
+
+    const dfs = (i: number) => {
+        if (vis.has(i)) return;
+
+        vis.add(i);
+        for (const j of g[i]) {
+            dfs(j);
+        }
+    };
+
+    for (const i of Object.keys(g)) {
+        if (vis.has(+i)) continue;
+
+        dfs(+i);
+        ans++;
+    }
```

Comment:
```suggestion
    for (let i = 0; i < n; i++) {
        if (!vis.has(i)) {
            dfs(i);
            ans++;
        }
    }
```

Reply:
From a code readability point of view, if (!vis.has(i)) { dfs(i); ans++; } might be better. This is because it explicitly expresses the logic of “if node i has not been visited, perform DFS and increase the number of connected components”, instead of using continue to skip nodes that have already been visited. This makes the logic of the code clearer and easier to understand.

Reply:
From the 1st glance yes, but... let't look from another point of view.
Every algorithm in most case is transformation input data into output.
And all the data can be considered as valid and invalid (simple binary choiсe, like in the "Matrix" movie))) And when invalid data rejected as early as possible common structure of the algorithm becomes more "flat", and cyclomatic and cognitive complexity is reduced. Also it is possible to draw an analogy with the domain of definition of a function from the school course of math. That is why usually I am trying to throw away wrong cases as early as possible. Also this approach is called as ["eager return"](https://greyfocus.com/2015/05/clean-code-pitfalls-eager-return/), and it shows itself in all its glory in the case of a large amount of code. 
And I am agree, in the current task it's advantages are not so obvious. But I am using it out of habbit over the years.
And also it make sense to add that you shouldn't make every good rule in absolut. :)

Reply:
I agree with you that `continue` is indeed pretty much the same whether to change it or not, there are pros and cons to each, and it's arguably a bit simpler to show if you don't use it over here. But Object.keys(g) needs to be changed, and I see that you've changed it as well.

The reason is as follows:

Performance: object.keys(g) creates a new array containing all the keys of g, which takes extra time and space. In contrast, for (let i = 0; i < n; i++) simply traverses a range, which is better performance.

Readability: for (let i = 0; i < n; i++) clearly expresses the range you want to traverse (from 0 to n-1), whereas const i of Object.keys(g) requires the reader to know that g is an object whose keys are the elements you want to traverse.

Type safety: in const i of Object.keys(g), i is a string, even though the keys of g are actually numbers. This can lead to type errors; for example, you may need to use +i to convert i to a number. In contrast, i in for (let i = 0; i < n; i++) is always a number, and there are no type problems.

Reply:
I am agree with you, and thank you for explanation.